src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1756,32 +1756,50 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
         public static int Min(this IEnumerable<int> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static int? Min(this IEnumerable<int?> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int? value = null;
-            foreach (int? x in source)
+            int? value;
+            using (IEnumerator<int?> e = source.GetEnumerator())
             {
-                if (value == null || x < value)
-                    value = x;
+                if (!e.MoveNext()) return null;
+                value = e.Current;
```

Comment:
Pre-initialising `value` goes against my instincts, but trying it does let us reduce the duplication down to a single `do…while` with pretty negliable effects (slight lose, but in the margin-of-error territory, and a win on empty sequences).
The `goto` isn't necessary. Indeed, I found it compiled to the same IL whether there was a `goto` or a `return`. No need to risk velociraptors.